src/frequenz/sdk/timeseries/logical_meter/_formula_engine.py

```diff
@@ -191,7 +120,7 @@ async def apply(self) -> Sample:
             raise RuntimeError("Some resampled metrics didn't arrive")
 
         if self._first_run:
-            metric_ts = await self.synchronize_metric_timestamps(ready_metrics)
+            metric_ts = await self._synchronize_metric_timestamps(ready_metrics)
```

Comment:
If user would run apply function twice without await, then it might happen that `synchronize_metric_timestamps` will be called twice.

Could you instead move setting _first_run here (before `await`):
```        
if self._first_run:
     self._first_run = False
     metric_ts = await self._synchronize_metric_timestamps(ready_metrics)
```

Reply:
We just have to make sure it gets run at least once.  Additional times is fine,  but the one time is essential.  So we can't assume it finished,  because it could raise an exception in rare cases as well.  We'll at some point start calling it once every `N` messages or minutes,  just to make sure the stream remains synchronized.

Reply:
Just FYI, off-topic but related, while writing the 2nd stage example, I also bumped into this issue (having to make sure I'm collecting samples from different sources but for the same resampling window), so I think it is worth pursuing #31 independently from the `FormulaEngine` (actually `FormulaEngine` could use it too).